catalyst/contrib/nn/modules/arcface.py

```diff
@@ -0,0 +1,229 @@
+        - Output: :math:`(batch, H_{out})` where
+          :math:`H_{out} = out\_features`.
+
+    Example:
+        >>> layer = SubCenterArcFace(5, 10, s=1.31, m=0.35, k=2)
+        >>> loss_fn = nn.CrosEntropyLoss()
+        >>> embedding = torch.randn(3, 5, requires_grad=True)
+        >>> target = torch.empty(3, dtype=torch.long).random_(5)
+        >>> output = layer(embedding, target)
+        >>> loss = loss_fn(output, target)
+        >>> loss.backward()
+
+    """
+
+    def __init__(  # noqa: D107
+        self,
+        in_features: int,
+        out_features: int,
+        s: float = 64.0,
+        m: float = 0.5,
+        k: int = 3,
+        eps: float = 1e-6,
+    ):
+        super(SubCenterArcFace, self).__init__()
+        self.in_features = in_features
+        self.out_features = out_features
+
+        self.s = s
+        self.m = m
+        self.k = k
+        self.eps = eps
+
+        self.weight = nn.Parameter(
+            torch.FloatTensor(k, in_features, out_features)
+        )
+        nn.init.xavier_uniform_(self.weight)
+
+        self.threshold = math.pi - self.m
+
+    def __repr__(self) -> str:
+        """Object representation."""
+        return (
+            "SubCenterArcFace("
+            + ",".join(
+                [
+                    f"in_features={self.in_features}",
+                    f"out_features={self.out_features}",
+                    f"s={self.s}",
+                    f"m={self.m}",
+                    f"k={self.k}",
+                    f"eps={self.eps}",
+                ]
+            )
+            + ")"
+        )
+
+    def forward(self, input, label):
+        """
+        Args:
+            input (torch.Tensor): input features,
+                expected shapes ``BxF`` where ``B``
+                is batch dimension and ``F`` is an
+                input feature dimension.
+            label (torch.Tensor): target classes,
+                expected shapes ``B`` where
+                ``B`` is batch dimension.
+
+        Returns:
+            tensor (logits) with shapes ``BxC``
+            where ``C`` is a number of classes.
+        """
+        cos_theta = torch.bmm(
+            F.normalize(input)
+            .unsqueeze(0)
+            .expand(self.k, *input.shape),  # k*b*f
+            F.normalize(
+                self.weight, dim=1
+            ),  # normalize in_features dim   # k*f*c
+        )  # k*b*f
+        cos_theta = torch.max(cos_theta, dim=0)[0]  # b*f
+        theta = torch.acos(
+            torch.clamp(cos_theta, -1.0 + self.eps, 1.0 - self.eps)
+        )
+
+        one_hot = torch.zeros(cos_theta.size()).to(input.device)
+        one_hot.scatter_(1, label.view(-1, 1).long(), 1)
+
+        selected = torch.where(
+            theta > self.threshold, torch.zeros_like(one_hot), one_hot
+        )
+
+        output = torch.cos(torch.where(selected.bool(), theta + self.m, theta))
+        output *= self.s
+        return output
```

Comment:
could you please rename `output` to `logits` for readability? In the docs you name it `logits`